src/hotspot/cpu/riscv/relocInfo_riscv.cpp

```diff
@@ -81,10 +79,9 @@ address Relocation::pd_call_destination(address orig_addr) {
 
 void Relocation::pd_set_call_destination(address x) {
   assert(is_call(), "should be an address instruction here");
-  if (MacroAssembler::is_call_at(addr())) {
-    address trampoline = nativeCall_at(addr())->get_trampoline();
-    if (trampoline != nullptr) {
-      nativeCall_at(addr())->set_destination_mt_safe(x, /* assert_lock */false);
+  if (NativeCall::is_at(addr())) {
+    NativeCall* nc = nativeCall_at(addr());
+    if (nc->reloc_set_destination(x)) {
```

Comment:
Seems there is a subtle difference here. Previously, there is a cache invalidation operation in `NativeCall::set_destination_mt_safe` [1]. Now it's gone with this change: it's not there in either `NativeShortCall::reloc_set_destination` or `NativeFarCall::reloc_set_destination`. Is that intended? 

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/nativeInst_riscv.cpp#L96

Reply:
Yes, we do an cache flush for the entire memory after all relocations are done.
```
void CodeBuffer::copy_code_to(CodeBlob* dest_blob) {
...
  relocate_code_to(&dest); // reloc done here
....
  // Flush generated code
  ICache::invalidate_range(dest_blob->code_begin(), dest_blob->code_size());
}
```

As we need todo this for the entire method I can't find a reason why reloc would need it?

Reply:
I.e. only in mt_safe case we need an invidual cache flush, if instructions where changed.

Reply:
Also I split them up because it was very confusing having relocations calling mt_safe.
As you can see in FarCall they are quite different, made no sense to having both cases in same method.

Reply:
Ah, I see. That make sense to me. Thanks.